iroh-net/src/magic_endpoint/rtt_actor.rs

```diff
@@ -0,0 +1,153 @@
+        /// Path changes for this connection from the magic socket.
+        conn_type_changes: ConnectionTypeStream,
+        /// For reporting-only, the Node ID of this connection.
+        node_id: NodeId,
+    },
+}
+
+/// Actor to coordinate congestion controller state with magic socket state.
+///
+/// The magic socket can change the underlying network path, between two nodes.  If we can
+/// inform the QUIC congestion controller of this event it will work much more efficiently.
+#[derive(Debug)]
+struct RttActor {
+    /// Stream of connection type changes.
+    connection_events: stream_group::Keyed<ConnectionTypeStream>,
+    /// References to the connections.
+    ///
+    /// These are weak references so not to keep the connections alive.  The key allows
+    /// removing the corresponding stream from `conn_type_changes`.
+    connections: HashMap<stream_group::Key, (quinn::WeakConnectionHandle, NodeId)>,
+    /// A way to notify the main actor loop to run over.
+    ///
+    /// E.g. when a new stream was added.
+    tick: Notify,
+}
+
+impl RttActor {
+    /// Runs the actor main loop.
+    ///
+    /// The main loop will finish when the sender is dropped.
+    async fn run(&mut self, mut msg_rx: mpsc::Receiver<RttMessage>) {
+        let mut cleanup_interval = tokio::time::interval(Duration::from_secs(5));
+        cleanup_interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
+        loop {
+            tokio::select! {
+                Some(msg) = msg_rx.recv() => self.handle_msg(msg),
+                item = self.connection_events.next(),
+                    if !self.connection_events.is_empty() => self.do_reset_rtt(item),
+                _ = cleanup_interval.tick() => self.do_connections_cleanup(),
+                () = self.tick.notified() => continue,
+                else => break,
+            }
+        }
+        debug!("rtt-actor finished");
+    }
+
+    /// Handle actor messages.
+    fn handle_msg(&mut self, msg: RttMessage) {
+        match msg {
+            RttMessage::NewConnection {
+                connection,
+                conn_type_changes,
+                node_id,
+            } => {
+                self.handle_new_connection(connection, conn_type_changes, node_id);
+            }
+        }
+    }
+
+    /// Handles the new connection message.
+    fn handle_new_connection(
+        &mut self,
+        connection: quinn::WeakConnectionHandle,
+        conn_type_changes: ConnectionTypeStream,
+        node_id: NodeId,
+    ) {
+        let key = self.connection_events.insert(conn_type_changes);
+        self.connections.insert(key, (connection, node_id));
+        self.tick.notify_one();
+    }
+
+    /// Performs the congestion controller reset for a magic socket path change.
+    fn do_reset_rtt(&mut self, item: Option<(stream_group::Key, ConnectionType)>) {
+        match item {
+            Some((key, new_conn_type)) => match self.connections.get(&key) {
```

Comment:
shouldn't this be checking the `new_conn_type`?

Reply:
I considered that, though I don't think this matters.  Every time the path changes we should reset this.  If we do direct -> mixed -> direct then we should fix the reason for this so it stay in direct (which we already did).  So I think keeping the logic here simple is preferred.

Reply:
maybe add a comment in this regard?